atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java

```diff
@@ -18,40 +18,56 @@
 
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.google.common.annotations.VisibleForTesting;
-import com.google.common.collect.ImmutableMap;
 import com.palantir.atlasdb.keyvalue.api.Cell;
 import com.palantir.atlasdb.keyvalue.api.CheckAndSetRequest;
 import com.palantir.atlasdb.keyvalue.api.KeyValueService;
 import com.palantir.atlasdb.keyvalue.api.TableReference;
 import com.palantir.atlasdb.keyvalue.api.Value;
 import com.palantir.atlasdb.schema.generated.TargetedSweepTableFactory;
+import com.palantir.atlasdb.sweep.asts.Bucket;
 import com.palantir.atlasdb.sweep.asts.bucketingthings.ObjectPersister.LogSafety;
+import com.palantir.atlasdb.sweep.queue.ShardAndStrategy;
+import com.palantir.common.streams.KeyedStream;
 import com.palantir.conjure.java.serialization.ObjectMappers;
+import com.palantir.logsafe.SafeArg;
 import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.Collection;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
 import java.util.function.Function;
+import java.util.stream.Collectors;
+
+final class DefaultSweepAssignedBucketStore implements SweepBucketAssignerStateMachineTable, SweepBucketPointerTable {
+    private static final SafeLogger log = SafeLoggerFactory.get(DefaultSweepAssignedBucketStore.class);
+    private static final int CAS_ATTEMPT_LIMIT = 5; // TODO: Remove this and share the logic somewhere.
 
-final class DefaultSweepAssignedBucketStore implements SweepBucketAssignerStateMachineTable {
     @VisibleForTesting
     static final TableReference TABLE_REF =
             TargetedSweepTableFactory.of().getSweepAssignedBucketsTable(null).getTableRef();
 
     private final KeyValueService keyValueService;
     private final ObjectPersister<BucketStateAndIdentifier> bucketStateAndIdentifierPersister;
+    private final ObjectPersister<Long> bucketIdentifierPersister;
 
     private DefaultSweepAssignedBucketStore(
             KeyValueService keyValueService,
-            ObjectPersister<BucketStateAndIdentifier> bucketStateAndIdentifierPersister) {
+            ObjectPersister<BucketStateAndIdentifier> bucketStateAndIdentifierPersister,
+            ObjectPersister<Long> bucketIdentifierPersister) {
         this.keyValueService = keyValueService;
         this.bucketStateAndIdentifierPersister = bucketStateAndIdentifierPersister;
+        this.bucketIdentifierPersister = bucketIdentifierPersister;
     }
 
     static DefaultSweepAssignedBucketStore create(KeyValueService keyValueService) {
         // We explicitly don't want to fail on unknown properties;
         ObjectMapper smileMapper = ObjectMappers.newClientSmileMapper();
```

Comment:
as mentioned in other PR: let's make sure this has a consistent ordering (whether here or elsewhere)